src/auth/googleauth.ts

```diff
@@ -123,182 +151,182 @@ export class GoogleAuth {
     // implemented yet)
 
     if (this._cachedProjectId) {
-      if (callback) setImmediate(callback, null, this._cachedProjectId);
-    } else {
-      const myCallback = (err?: Error|null, projectId?: string) => {
-        if (!err && projectId) {
-          this._cachedProjectId = projectId;
-        }
-        if (callback) setImmediate(callback, err, projectId);
-      };
+      return this._cachedProjectId;
+    }
 
-      // environment variable
-      if (this._getProductionProjectId(myCallback)) {
-        return;
-      }
+    let projectId: string|null|undefined;
 
-      // json file
-      this.getFileProjectId((err?: Error|null, projectId?: string) => {
-        if (err || projectId) {
-          myCallback(err, projectId);
-          return;
-        }
+    // environment variable
+    projectId = this.getProductionProjectId();
+    if (projectId) {
+      this._cachedProjectId = projectId;
+      return projectId;
+    }
 
-        // Google Cloud SDK default project id
-        this.getDefaultServiceProjectId(
-            (err2?: Error|null, projectId2?: string) => {
-              if (err2 || projectId2) {
-                myCallback(err2, projectId2);
-                return;
-              }
-
-              // Get project ID from Compute Engine metadata server
-              this.getGCEProjectId(myCallback);
-            });
-      });
+    // json file
+    projectId = await this.getFileProjectId();
+    if (projectId) {
+      this._cachedProjectId = projectId;
+      return projectId;
     }
+
+    // Google Cloud SDK default project id
+    projectId = await this.getDefaultServiceProjectId();
+    if (projectId) {
+      this._cachedProjectId = projectId;
+      return projectId;
+    }
+
+    // Get project ID from Compute Engine metadata server
+    projectId = await this.getGCEProjectId();
+    if (projectId) {
+      this._cachedProjectId = projectId;
+      return projectId;
+    }
+
+    return null;
   }
 
   /**
    * Run the Google Cloud SDK command that prints the default project ID
-   * @param {function} callback Callback.
-   * @api private
    */
-  _getSDKDefaultProjectId(
-      callback:
-          (error: Error|null, stdout: string, stderr: string|null) => void) {
-    exec('gcloud -q config list core/project --format=json', callback);
+  _getSDKDefaultProjectId():
```

Comment:
I think this is private, but expose for testing? If so could this be an async function so that you don't have to manual promising below (nit).

Reply:
It is private, but exposed for testing. If I make this an async, I need to promisify exec.  I think I avoided doing that so I wouldn't need to pull in pify just for this one case.  What do you think?  Is it worth adding another dependency to clean this up?

Reply:
Add a TODO.

Reply:
Done